client/src/ts/dial.ts

```diff
@@ -0,0 +1,206 @@
+export enum LineEnd {
+  Flat = 1,
+  Round = 2,
+}
+
+// animate the curve of the arc using a cubic easing function and animate
+// frames.
+function Animate(start: number, end: number, duration: number, step: number,
+                 callback: (new_val: number) => void): void {
+  let currentIteration = 1;
+  const interations = 60 * duration;
+  if (start > end) {
+    step = step * -1;
+  }
+
+  function easeCubic(pos: number): number {
+    pos /= 0.5;
+    if (pos < 1) {
+      return 0.5 * Math.pow(pos, 3);
+    }
+    return 0.5 * (Math.pow(pos - 2, 3) + 2);
+  }
+
+  function animate(): void {
+    const progress = currentIteration++ / interations;
+    const value = start + step * currentIteration * easeCubic(progress);
+    callback(Math.round(value));
+    if (step > 0 && value < end) {
+      window.requestAnimationFrame(animate);
+    } else if (step < 0 && value > end) {
+      window.requestAnimationFrame(animate);
+    }
+  }
+
+  window.requestAnimationFrame(animate);
+}
+
+// maps polar coordinates to Cartesian to draw the dial.
+function polarToCartesian(centerX: number, centerY: number, radius: number,
+                          angle: number): any {
+  return {
+    x : centerX + radius * Math.cos(angle),
+    y : centerY + radius * Math.sin(angle)
+  };
+}
+
+// describes the SVG arc as a string.
+function describeArc(x: number, y: number, radius: number, startAngle: number,
+                     endAngle: number, reverse: boolean): string {
+  const start = polarToCartesian(x, y, radius, endAngle);
+  const end = polarToCartesian(x, y, radius, startAngle);
+  const largeArcFlag = endAngle - startAngle <= Math.PI ? '0' : '1';
+
+  return [
+    'M', start.x, start.y, 'A', radius, radius, 0, largeArcFlag, 0, end.x, end.y
+  ].join(' ');
+}
+
+// options to adjust the behavior of Dial.
+export class DialOptions {
+  // shape
+  public min: number = 0;
+  public max: number = 100;
+  public step: number = 1;
+  public angleOffset: number = 0;
+  public angleArc: number = 2 * Math.PI;
+  public rotation: Direction = Direction.Clockwise;
+
+  // ui
+  public duration: number = 1;
+  public thickness: number = 1;
+  public padding: number = 5;
+  public lineCap: LineEnd = LineEnd.Flat;
+  public dialColor: string = '#000000';
+  public dialOpacity: number = 1;
+  public rectColor: string = '#ffffff';
+  public rectOpacity: number = 1;
+  public showText: boolean = true;
+  public font: string = 'Open Sans';
+  public fontSize: number = 30;
+  public fontColor: string = '#000000';
+}
+
+// class for drawing SVG based dials.
+export class Dial {
+  private _svg: SVGElement;
+  private _rect: SVGRectElement;
+  private _path: SVGPathElement;
+  private _text: SVGTextElement;
+  private _value: number = 0;
+  private _delta: number;
+  private _div: HTMLDivElement;
+  private _options: DialOptions;
+
+  constructor(div: HTMLDivElement, options: DialOptions) {
+    this._div = div;
+    this._options = options;
```

Comment:
What's the difference between setting this here, and when you call `this.updateOptions(options)`?